exams/tasks.py

```diff
@@ -2,56 +2,130 @@
+    """
+    return settings.EXAMS_SFTP_BACKOFF_BASE ** retries
+
+
+@async.task(bind=True, max_retries=3)
+def export_exam_profiles(self):
     """
     Sync any outstanding profiles
     """
-    exam_profiles = (ExamProfile.objects
-                     .filter(status=ExamProfile.PROFILE_PENDING)
-                     .select_related('profile'))
-    file_prefix = datetime.now(pytz.utc).strftime('cdd-%Y%m%d%H_')
+    exam_profiles = ExamProfile.objects.filter(
+        status=ExamProfile.PROFILE_PENDING).select_related('profile')
+    file_prefix = datetime.now(pytz.utc).strftime(PEARSON_CDD_FILE_PREFIX)
 
     # write the file out locally
     # this will be written out to a file like: /tmp/cdd-20160405_kjfiamdf.dat
     with tempfile.NamedTemporaryFile(
         prefix=file_prefix,
-        encoding='utf-8',
-        suffix='.dat',
+        encoding=PEARSON_FILE_ENCODING,
+        suffix=PEARSON_FILE_EXTENSION,
         mode='w',
     ) as tsv:
-        valid_profiles, invalid_profiles = write_cdd_file(tsv, exam_profiles)
+        cdd_writer = pearson.CDDWriter()
+        valid_profiles, invalid_profiles = cdd_writer.write(tsv, exam_profiles)
 
         # flush data to disk before upload
         tsv.flush()
 
-        # upload to SFTP server
-        upload_tsv(tsv.name)
+        try:
+            # upload to SFTP server
+            pearson.upload_tsv(tsv.name)
+        except RetryableSFTPException as exc:
+            log.exception('Retryable error during upload of CDD file to Pearson SFTP')
+            # retry up to 3 times w/ exponential backoff if this was a connection error
+            self.retry(exc=exc, countdown=_backoff(self.request.retries))
+        except:  # pylint: disable=bare-except
+            log.exception('Unexpected exception uploading CDD file')
+            return
+
+    valid_profile_ids = [exam_profile.id for exam_profile in valid_profiles]
+    invalid_profile_ids = [exam_profile.id for exam_profile in invalid_profiles]
 
     with transaction.atomic():
-        # update records to reflect the successful upload
-        if valid_profiles:
-            exam_profile_ids = [exam_profile.id for exam_profile in valid_profiles]
-            (ExamProfile.objects
-             .filter(id__in=exam_profile_ids)
-             .update(status=ExamProfile.PROFILE_IN_PROGRESS))
-
-        # update records to reflect invalid profile
-        if invalid_profiles:
-            exam_profile_ids = [exam_profile.id for exam_profile in invalid_profiles]
-            (ExamProfile.objects
-             .filter(id__in=exam_profile_ids)
-             .update(status=ExamProfile.PROFILE_INVALID))
+        try:
+            # update records to reflect the successful upload
+            if valid_profile_ids:
+                ExamProfile.objects.filter(
+                    id__in=valid_profile_ids).update(status=ExamProfile.PROFILE_IN_PROGRESS)
+
+            # update records to reflect invalid profile
+            if invalid_profile_ids:
```

Comment:
Yes they could - I'm not quite sure what you're alluding to but now that I'm reading my implementation here again, this `try..except` inside the atomic transaction is not a good thing - it hides from Django that the transaction broke. I'd propose wrapping the entire transaction block instead. If you feel we need per-query logging on which one errored that we don't otherwise get in the traceback I can also wrap/log/reraise each query.